### PR TITLE
Stream GRPO log probabilities

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Moves GRPO log-probabilities and entropy onto `_ChunkedLogProbFunction`, then runs the trainer's native loss unchanged. This enables off-policy and entropy options and removes the duplicate GRPO loss files. MoE auxiliary loss keeps the regular model forward.

Part of #7063.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. @qgallouedec

---

> [!NOTE]
> **Medium Risk**
> Changes the `use_liger_kernel` GRPO training path by removing the fused loss and routing all objectives through `_compute_loss`; behavior should match but this touches core RL optimization and distributed LM-head handling.
> 
> **Overview**
> **GRPO with `use_liger_kernel=True` no longer uses a separate fused Liger GRPO loss.** Log-probs and entropies are computed via the shared streamed `_ChunkedLogProbFunction` (backbone + direct `lm_head` projection), then the same `_compute_loss` path as the non-Liger trainer runs. That unifies behavior across loss types and drops duplicate code: `trl/losses/grpo_loss.py`, `trl/losses/fused_linear_ppo.py`, and `tests/test_grpo_loss.py` are removed, and `FusedLinearGRPOLoss` is no longer exported.
> 
> The chunked path adds **autocast**, **FSDP DTensor head gathering**, **`logit_scale` / `final_logit_softcapping`**, and **`masked_scatter`** so all-masked completions still backprop a differentiable zero. **MoE aux loss** still uses the full forward when requested. Several **Liger-only guardrails are removed** (off-policy masking, entropy bonus/filter, sequence importance sampling), and docs drop the fused-kernel-only `clip_ratio` metric.
> 
> Tests shift from Liger-vs-reference loss parity to **chunked vs full-logits logprob parity**, bf16 projection checks, reference scoring, MoE training, and broader `use_liger_kernel` coverage on existing trainer tests.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e69fcee96e80dd4ae88a10754fe9992f7a9c4e25. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1525,29 +1457,66 @@ def _get_per_token_logps_and_entropies(
...
+                labels = completion_ids[completion_mask]
+                lm_head = inner_model.get_output_embeddings()
+                text_config = inner_model.config.get_text_config()
+                final_logit_softcapping = getattr(text_config, "final_logit_softcapping", None)
+                logit_scale = getattr(text_config, "logit_scale", 1.0)
+                with maybe_gather_lm_head_ctx(lm_head.weight, lm_head.bias):
+                    valid_logps, valid_entropies = _ChunkedLogProbFunction.apply(
+                        hidden_states,
+                        lm_head.weight,
+                        lm_head.bias,
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/grpo_trainer.py]: Fixed in 182c27e5. The FSDP2 head weight and bias are materialized once before the token chunks, following the chunked SFT path.

[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/grpo_trainer.py]: Fixed in 182c27e5. Backward now casts each weight chunk to the hidden-state dtype just like forward, with a mixed bf16/fp32 regression test.

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1525,29 +1457,66 @@ def _get_per_token_logps_and_entropies(
...
+                hidden_states = outputs.last_hidden_state[:, :-1]
+                hidden_states = hidden_states[:, -logits_to_keep:]
+                completion_mask = attention_mask_batch[:, -logits_to_keep:].bool()
+
+                hidden_states = hidden_states[completion_mask]
+                labels = completion_ids[completion_mask]
+                lm_head = inner_model.get_output_embeddings()
+                text_config = inner_model.config.get_text_config()
+                final_logit_softcapping = getattr(text_config, "final_logit_softcapping", None)
+                logit_scale = getattr(text_config, "logit_scale", 1.0)
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/grpo_trainer.py]: Fixed in 182c27e5. `None` is normalized to 1.0 and `output_multiplier` is used as the fallback, matching chunked SFT. The regression test covers both.

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -118,6 +118,9 @@
 
 logger = get_logger(__name__)
 
+
+_CHUNKED_LOGPROB_CHUNK_SIZE = 8192
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/grpo_trainer.py:122]: Fixed in 182c27e5. Valid completion tokens are now split into 2,048-row chunks, retaining the bounded token side of the previous dual-chunked implementation.

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -1525,29 +1457,66 @@ def _get_per_token_logps_and_entropies(
...
+                        logit_scale,
+                    )
+
+                logps = torch.zeros_like(completion_mask, dtype=valid_logps.dtype)
+                logps[completion_mask] = valid_logps
+                all_logps.append(logps)
+                if compute_entropy:
+                    entropies = torch.zeros_like(completion_mask, dtype=valid_entropies.dtype)
+                    entropies[completion_mask] = valid_entropies
+                    all_entropies.append(entropies)
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/grpo_trainer.py:1512]: The two paths implement the same selected-token log-probabilities, so using the streamed path for old/reference scores does not bias the ratio. `test_chunked_loss_matches_full_logits` compares loss and every model gradient across all supported loss types (including KL), and the new scaling regression covers model-specific logit transforms. Forcing these scoring calls through full logits would restore the memory peak this PR removes; the causal-LM path remains only where the MoE wrapper is needed to return `aux_loss`.

**Comment by kashif** [comment]
@codex review

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -1525,29 +1458,92 @@ def _get_per_token_logps_and_entropies(
...
-                if self._entropy_bonus_enabled:
-                    entropies = entropy_from_logits(logits)
+            if self.use_liger_kernel and not compute_aux_loss:
+                inner_model = model.base_model.model if is_peft_model(model) else model
+                if self._is_vlm and Version(transformers.__version__) < Version("5.0.0"):
+                    backbone = inner_model.model
                 else:
-                    with torch.no_grad():
+                    backbone = inner_model.base_model
+                outputs = backbone(**model_inputs)
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/grpo_trainer.py]: Fixed in 071b2752. Direct backbone scoring now re-enters Accelerate autocast, and the projection uses the active autocast dtype too. The regression checks an actual Q projection runs in bf16 and the resulting log-probs stay finite.

**Comment by kashif** [comment]
@codex review

**Comment by kashif** [comment]
@codex review

File: trl/trainer/utils.py [hunk 6]
```diff
@@ -1339,58 +1342,76 @@ def forward(
...
+            target_logit = torch.zeros((n_chunk,), device=device, dtype=torch.float32)
+            row_idx = torch.arange(n_chunk, device=device)
+
+            for start in range(0, vocab, chunk_size):
+                end = min(start + chunk_size, vocab)
+                C = end - start
+                # using fp16=True, the model's hidden states get cast to float16 by autocast, but the mm_buf is
+                # allocated with last_hidden.dtype (float16) while w_chunk (the lm_head weights) is not auto casted
+                w_chunk = weight[start:end].to(compute_dtype)  # [C, H]
+                torch.mm(hidden_chunk.to(compute_dtype), w_chunk.t(), out=mm_buf[:n_chunk, :C])
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/utils.py:1378]: This matches the non-chunked path. Under Accelerate bf16 autocast, the normal `lm_head.forward` projection returns bf16 even when `cast_lm_head_to_fp32=True` and the stored head weight is fp32 (verified with tiny Qwen3: fp32 weight, bf16 full logits). Autocast chooses the projection compute dtype; the flag preserves fp32 parameter storage/updates. The chunked mixed-precision test checks the same bf16 compute behavior, while the separate mixed bf16-hidden/fp32-weight unit test covers backward.

**Comment by kashif** [comment]
@codex review

**Comment by kashif** [comment]
@codex review

**Comment by qgallouedec** [comment]
Nice!

Four 3 things at this point.

- It reverts #7076. No `ctx.needs_input_grad` in your `backward`, all three fp32 buffers unconditional again. #7076 lands first per #7063, rebasing must keep it.

- The token chunking is untested: you add a token loop at 2048, but the test runs bs=3, 3 generations, 32 completion tokens, so N maxes at 96 and the loop body runs once. (Vocab loop is fine, 152064/8192 = 19 iterations.). Add a case above 2048?

- `grad_atol = 2e-2` for luspo/vespo.** 40x the 5e-4 everywhere else, on absolute gradients, fp32, tiny model. I get the mechanism, both exponentiate a sequence-level sum so per-token error amplifies, but then bound the amplification or tighten the logprob parity. A 40x gate accepts a real discrepancy in exactly the two loss types with the most delicate math. Same reason I reverted f527bc3f.

**Comment by kashif** [comment]
Rebased on main and fixed the conflict.

- The chunked backward now gates the three gradient buffers on `ctx.needs_input_grad` again, same as #7076.
- Added `test_chunked_loss_matches_full_logits_multiple_token_chunks`: 3*700=2100 completion tokens so the token loop runs twice, not once.
- `grad_atol` for luspo/vespo is now `5e-4 * max_completion_length` instead of a flat 2e-2, so it scales with the thing that amplifies the error instead of being a magic constant.

File: trl/trainer/grpo_trainer.py [hunk 7]
```diff
@@ -1449,7 +1364,25 @@ def get_high_entropy_mask(self, entropies: torch.Tensor, mask: torch.Tensor, thr
...
+                # Scoring runs outside `compute_loss`, so enter through distributed wrappers before reading the
+                # backbone and LM head directly.
+                return self._forward_redirection(
+                    model,
+                    unwrapped_model,
+                    self._get_per_token_logps_and_entropies_impl,
+                    unwrapped_model,
+                    *args,
+                    **kwargs,
+                )
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/grpo_trainer.py:1381]: Fixed in 99582873. `_compute_loss` now calls `_get_per_token_logps_and_entropies_impl` directly instead of the public wrapper, since `compute_loss` already redirected `model` through the distributed wrapper before calling `_compute_loss` — the nested call was re-redirecting on an already-unwrapped object. Traced through `_ForwardRedirection`: when wrapper and original are the same object it doesn't crash (it degrades to a plain __call__), so this wasn't corrupting training, just pointless indirection, but it's cleaner to not do it. I don't have multi-GPU FSDP hardware to exercise test_grpo_liger in tests/distributed/test_distributed.py here, so this is verified by code tracing plus the single-GPU liger/chunked suite (unaffected), not by running the actual FSDP2 case.

File: trl/trainer/grpo_trainer.py [hunk 8]
```diff
@@ -3120,7 +3042,10 @@ def _compute_loss(self, model, inputs):
         mask = completion_mask if "tool_mask" not in inputs else completion_mask * inputs["tool_mask"]
 
         # Compute the per_token_logps and the entropy at each position in the completion
-        per_token_logps, entropies, aux_loss = self._get_per_token_logps_and_entropies(
+        # `compute_loss` already redirected `model` through the distributed wrapper before calling us, so call the
+        # implementation directly instead of `_get_per_token_logps_and_entropies`, which would try to redirect again
+        # on an already-unwrapped model.
+        per_token_logps, entropies, aux_loss = self._get_per_token_logps_and_entropies_impl(
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/grpo_trainer.py]: Reverted in 583fde8. `MiniLLMTrainer.compute_loss` overrides the base and calls `_compute_loss` directly with the still-wrapped model, skipping the redirect entirely — calling `_get_per_token_logps_and_entropies_impl` unconditionally broke that path under FSDP. Good catch. The redundant redirect this was trying to avoid is harmless (`_ForwardRedirection` degrades to a plain `__call__` when wrapper and original are the same object), so keeping it is the safer choice over trying to special-case every caller.

File: trl/trainer/grpo_trainer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/grpo_trainer.py:1381]: Correction: reverted that fix in 583fde8 — it broke MiniLLMTrainer's direct call path (see the follow-up bugbot finding on this same commit). Leaving this as-is: the double redirect is wasteful but, per the trace above, not actually harmful (same-object wrapper/original degrades to a plain __call__, no re-entrant FSDP hook firing).

File: trl/trainer/grpo_trainer.py [hunk 9]
```diff
@@ -1532,29 +1464,82 @@ def _get_per_token_logps_and_entropies(
...
-            logps = selective_log_softmax(logits, completion_ids)  # compute logprobs
-            all_logps.append(logps)
-
-            if compute_entropy:
-                # The entropy bonus is a differentiable loss term, so entropies must carry grad when it is
-                # active. Otherwise they only feed logging and the top_entropy_quantile mask (neither
-                # differentiable), so we skip grad to avoid retaining the memory-heavy full-vocab softmax.
-                if self._entropy_bonus_enabled:
-                    entropies = entropy_from_logits(logits)
+            if self.use_liger_kernel and not compute_aux_loss:
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1468]: This all runs inside the batch loop:

```python
for start in range(0, input_ids.size(0), batch_size):
```

The loop is there to cut the `lm_head` peak. The chunked head already does that, over both vocab and tokens, so the liger branch doesn't need it.

I'd move the whole liger branch outside the loop. The loop then wraps the non-liger path only, and disappears with it when we go liger-only.

**Comment by qgallouedec** [comment]
LGTM!!
One leftover for whenever the three chunked PRs settle: `_CHUNKED_LOGPROB_CHUNK_SIZE` is defined in `grpo_trainer.py` while `_CHUNKED_LOGPROB_TOKEN_CHUNK_SIZE` sits in `utils.py`, and #7243 adds a third copy. Worth putting both in one place at some point

**Comment by qgallouedec** [comment]
Pushed three test-only commits on top of yours, net -146 lines.

- **Parity**: the 9 loss-type cases and the multi-token-chunk test collapse into one `test_chunked_logps_match_full_logits`, helper gone. Loss math is path-independent, so matching log-probs covers every loss type. Comparing at that level also drops the dataloader dance, `current_gradient_accumulation_steps` and `_metrics.clear()`. Patched LM head kept, and 2200 tokens still crosses a token chunk.
- **Scaling**: nothing tested `logit_scale` or `final_logit_softcapping`, so that moves to `TestChunkedLogProbFunction`. Dropped the `output_multiplier` half, it only covered a `getattr` fallback.
- **Empty completions**: now `use_liger_kernel` on the existing `test_train_with_mask_truncated_completions_all_masked`. KTO has no equivalent config, so it keeps a direct test.
- **Mixed precision**: hooks `q_proj` across `train()` instead of calling `_get_per_token_logps_and_entropies` by hand.
- Same mirrored in KTO, plus the missing `# reduce ...` comments.

One thing I couldn't align: GRPO branches on `use_liger_kernel` inside `_get_per_token_logps_and_entropies`, KTO at each call site, so KTO's parity test stays at loss level. Worth unifying separately, along with KTO returning `outputs` where GRPO returns `aux_loss`.

**Comment by qgallouedec** [comment]
Restructured `_get_per_token_logps_and_entropies`. It wrapped a second method that branched on `use_liger_kernel` and inlined both implementations, so there were three layers and two of them only forwarded arguments.

Now a dispatcher over `_chunked_logps` and `_full_logits_logps` (will be way easier to remove in the future), with the redirection next to the only path that needs it: the chunked one reads the backbone and LM head directly, the full-logits one calls the model and needs no redirect.

`_chunked_logps` drops `batch_size` and `compute_aux_loss`, neither of which applies to it. That surfaced a call site passing `batch_size` positionally, where it would have landed on `compute_entropy`: entropy computed unnecessarily and the batch size never applied. Now passed by keyword like the other two.

One thing to check before merging: the MoE plus liger path no longer goes through `_forward_redirection`, since it takes the full-logits branch anyway. I'm working on aux loss concurrently anyway.

**Review by qgallouedec — APPROVED** [review]
thanks!

File: trl/trainer/grpo_trainer.py [hunk 10]
```diff
@@ -796,40 +799,34 @@ def get_reward(environments, _env_type=env_type, **kwargs):
...
                     "wrong sequence. Use a weight-based adapter such as LoRA instead, or set "
                     "`use_liger_kernel=False`."
                 )
         self.mask_truncated_completions = args.mask_truncated_completions
         self.top_entropy_quantile = args.top_entropy_quantile
-        if self.use_liger_kernel and self.top_entropy_quantile < 1.0:
-            raise NotImplementedError(
-                "Liger Kernels don't currently support masking token positions based on entropy."
-            )
-        if self.use_liger_kernel and self.importance_sampling_level not in ("token", "sequence"):
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:832]: ah right! GSPO-token overrides `_compute_loss` and handles `sequence_token` there, so the init check was the only thing rejecting it. Dropped the check: base GRPO already rejects unknown levels in `_compute_loss`, and subclasses can extend the set.

**Comment by qgallouedec** [comment]
Not sure why min is failing, merging so that we can move forward, i will investigate separately

**Comment by DimensionSTP** [comment]
Hi @kashif, #7077 also superseded the production fix in #6699. I've adapted that PR into a test-only follow-up for the new chunked log-prob path, preserving the Gemma 3 and Qwen3.5 + PEFT regression coverage.

The updated suite passed locally on CUDA with no skips. Would this coverage be useful to retain? I'd appreciate your thoughts on #6699:
